boa/src/builtins/regexp/mod.rs

```diff
@@ -188,114 +185,210 @@ impl RegExp {
-            unicode = true;
-            sorted_flags.push('u');
-        }
-        if regex_flags.contains('y') {
-            sticky = true;
-            sorted_flags.push('y');
+        let mut sticky = false;
+        for c in f.chars() {
+            match c {
+                'g' if global => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 'g'")
+                }
+                'g' => global = true,
+                'i' if ignore_case => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 'i'")
+                }
+                'i' => ignore_case = true,
+                'm' if multiline => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 'm'")
+                }
+                'm' => multiline = true,
+                's' if dot_all => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 's'")
+                }
+                's' => dot_all = true,
+                'u' if unicode => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 'u'")
+                }
+                'u' => unicode = true,
+                'y' if sticky => {
+                    return context.throw_syntax_error("RegExp flags contains multiple 'y'")
+                }
+                'y' => sticky = true,
+                c => {
+                    return context.throw_syntax_error(format!(
+                        "RegExp flags contains unknown code unit '{}'",
+                        c
+                    ))
+                }
+            }
         }
 
-        let matcher = match Regex::with_flags(&regex_body, sorted_flags.as_str()) {
+        // 12. Set obj.[[OriginalSource]] to P.
+        // 13. Set obj.[[OriginalFlags]] to F.
+        // 14. Set obj.[[RegExpMatcher]] to the Abstract Closure that evaluates parseResult by applying the semantics provided in 22.2.2 using patternCharacters as the pattern's List of SourceCharacter values and F as the flag parameters.
+        let matcher = match Regex::with_flags(&p, f.as_ref()) {
             Err(error) => {
-                return Err(
-                    ctx.construct_syntax_error(format!("failed to create matcher: {}", error.text))
-                );
+                return Err(context
+                    .construct_syntax_error(format!("failed to create matcher: {}", error.text)));
             }
             Ok(val) => val,
         };
 
         let regexp = RegExp {
             matcher,
             use_last_index: global || sticky,
-            flags: sorted_flags.into_boxed_str(),
             dot_all,
             global,
             ignore_case,
             multiline,
             sticky,
             unicode,
-            original_source: regex_body,
-            original_flags: regex_flags,
+            original_source: p,
+            original_flags: f,
```

Comment:
I think its best to make `original_source` and `original_flags` `JsString`s

Reply:
Done in a0d9713.

No changes in the tests expect for `test/built-ins/Map/prototype/keys/returns-iterator.js`. But I think that one is just panicking randomly sometimes.